server/blob/blob-cloud/src/main/java/org/apache/james/blob/cloud/CloudBlobsDAO.java

```diff
@@ -0,0 +1,120 @@
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.cloud;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.HashBlobId;
+import org.apache.james.blob.api.ObjectStore;
+import org.apache.james.blob.api.ObjectStoreException;
+import org.jclouds.ContextBuilder;
+import org.jclouds.blobstore.BlobStore;
+import org.jclouds.blobstore.domain.Blob;
+import org.jclouds.blobstore.options.CopyOptions;
+import org.jclouds.logging.slf4j.config.SLF4JLoggingModule;
+import org.jclouds.openstack.swift.v1.blobstore.RegionScopedBlobStoreContext;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.hash.Hashing;
+import com.google.common.hash.HashingInputStream;
+import com.google.inject.Inject;
+import com.google.inject.Module;
+
+class CloudBlobsDAO implements ObjectStore {
+    private static final InputStream EMPTY_STREAM = new ByteArrayInputStream(new byte[0]);
+    private static final Iterable<Module> JCLOUDS_MODULES = ImmutableSet.of(new SLF4JLoggingModule());
+
+    private final BlobId.Factory blobIdFactory;
+    private final BlobStore blobStore;
+    private final ContainerName containerName;
+
+    @Inject
+    public CloudBlobsDAO(ContainerName containerName, HashBlobId.Factory blobIdFactory,
+                         CloudsBlobsConfiguration cloudsBlobsConfiguration) {
+        this.blobIdFactory = blobIdFactory;
+        this.containerName = containerName;
+
+        RegionScopedBlobStoreContext blobStoreContext = ContextBuilder.newBuilder("openstack-swift")
+            .endpoint(cloudsBlobsConfiguration.getEndpoint().toString())
```

Comment:
we never rely on toString methods when we can avoid it (toString is supposed to be for debug representation of the object)

Reply:
I couldn't find another method on `java.net.URI` to get the string representation of the URI. I didn't want to go through `java.net.URL` as it does network access on the most unexpected calls. 
The alternatives I could think of are :
- writing my own Function1<URI,String> but that's not trivial
- writing my own URI/URL representation  which is not trivial either
- import a 3rd party library to get an alternative URI/URL representation (I didn't research that)

Reply:
oh, sorry, I thought this `getEndpoint()` returned something you wrote. URI is a known case where we actually call toString. No problem, keep it that way.

Reply:
should I rename it getEndpointURI ?

Reply:
No, that's ok, I don't really like encoding types in method names